### ISSUE TITLE
How to debug 'Unpaired reads when require_paired is set!' ?

### ISSUE THREAD

**Issue by bede** [dib-lab/khmer#1338]
I have a some interleaved reads which `normalize-by-median.py` doesn't like in paired end mode. They have been interleaved using `interleave-reads.py`  
Command:
```
normalize-by-median.py -C {c} -k {k} -N 4 -x 1e9 -p  {out}/trim/{name}.fr.fastq
```
Error:
```
ERROR: Unpaired reads when require_paired is set!
```
I've put these reads them through a paired read validation script and they pass. A small subset of the reads works fine. Is there a way to get Screed to report the culprit line(s)? How do I debug this issue? I've set the logging level to `INFO` but it still doesn't provide any helpful info.

Below are the first two reads from one such file, but running just these two reads *doesn't* cause the crash:
```
@M02383:22:000000000-A9G0P:1:1101:14743:1579 1:N:0:4
CATCAATTCAACTCTTCAGGTTGCCCCGAGAGGATGGCCAGCTGCAGACCCCTCGCCGATTTCGATCAAGGCTGGGGCCCTATCAGCTACGCCAACGGAAGCGGCCCCGAACACCGGCCCTACTGCTGGCACTATCCTCCAAAGCCTTGTG
+
1>AAADFFFFFFGGGGG3BFGGCFHHFEGEAA0BGEAEGHHGHHHFFEHGHGGGGGGGGGGGHFFGHFHHHHHHGGGGGGHHHHHHHHHHGGGGGGEGGGEGCGGGGGGGCCCGEGCGGGHHHHGHHGHGHHHHHHHHHHHHEHHHHHHHH
@M02383:22:000000000-A9G0P:1:1101:14743:1579 2:N:0:4
CACCCCTCCGATGACACAAGGAGGCGCTCCGCACACTTTGGTACATCCAGATGAGTTCATCCAGGTGCAACCGAACCAATTGCCCAACGGTGGCCTGGTGTTGTTGAGGACGAAGACGTCCGTATCATTCTCACCCCAGTTGTAGGTAGG
+
BBBBBBABBABBEBBFGGGBABBEFGGGGGFGGGGHHHHBFBD55DFGEGHHHEGHHHHHHHGFBFHHG@GHGGCGGHFHHHGHHHHGGGGHGHHHGGHHHGGHHHHHHHHG?FGGHGGGGGGHFHHHHHHHHHHHGGGHHGHHGHHB1E
```

**Comment by ctb** [comment]
On Tue, Jan 19, 2016 at 06:50:18AM -0800, Bede Constantinides wrote:
> I have a some interleaved reads which `normalize-by-median.py` doesn't like in paired end mode. They have been interleaved using `interleave-reads.py`  
> Command:
> ```
> normalize-by-median.py -C {c} -k {k} -N 4 -x 1e9 -p  {out}/trim/{name}.fr.fastq
> ```
> Error:
> ```
> ERROR: Unpaired reads when require_paired is set!
> ```
> I've put these reads them through a paired read validation script and they pass. A small subset of the reads works fine. Is there a way to get Screed to report the culprit line(s)? How do I debug this issue? I've set the logging level to `INFO` but it still doesn't provide any helpful info.

yes, something I've seen enough times myself to want to provide more debugging
info :)

I can give you some python code to run... what do you think would be useful
output? Just the names of the reads?

thx

**Comment by bede** [comment]
Woah that was fast! Yes please, do share @ctb

**Comment by ctb** [comment]
```
#! /usr/bin/env python
import sys
import screed
from khmer.utils import broken_paired_reader

seqiter = screed.open(sys.argv[1])

for n, is_pair, read1, read2 in broken_paired_reader(seqiter):
   if not is_pair:
      print read1.name
```

should print the offending reads - you can add 'require_paired=True'
to broken_paired_reader to have it raise an exception, and/or just
exit after the first print, too.

HTH! And let me know where the bug is... <sigh>

Come to think of it - I'm guessing it would be useful (to me, if no one else)
if khmer were to save the offending context to a file.  Hmm.

**Comment by bede** [comment]
Thanks Titus! I'm not seeing any reads reported by your script...

Any ideas?

normalize-by is definitely failing

```
l-umobcdja2:xx Bede$ normalize-by-median.py -C 1 -k 21 -N 4 -x 1e9 -p 6.fr.fastq

|| This is the script normalize-by-median.py in khmer.
|| You are running khmer version 2.0
|| You are also using screed version 0.9
||
|| If you use this script in a publication, please cite EACH of the following:
||
||   * MR Crusoe et al., 2015. http://dx.doi.org/10.12688/f1000research.6924.1
||   * CT Brown et al., arXiv:1203.4802 [q-bio.GN]
||
|| Please see http://khmer.readthedocs.org/en/latest/citations.html for details.


PARAMETERS:
 - kmer size =    21 		(-k)
 - n tables =     4 		(-N)
 - max tablesize = 1e+09 	(-x)

Estimated memory usage is 4e+09 bytes (n_tables x max_tablesize)
--------
making countgraph
** ERROR: Unpaired reads when require_paired is set!
** Failed on 6.f.fr.fastq: 
** Exiting!
l-umobcdja2:xx Bede$
```

**Comment by ctb** [comment]
On Tue, Jan 19, 2016 at 07:17:56AM -0800, Bede Constantinides wrote:
> Thanks Titus! I'm not seeing any reads reported by your script...

OK - one last thing to try. There's a bug in broken_paired_reader
where, if one sequence is shorter than k in length, it's ignored;
this triggers complaints about broken pairing.  Could it be that
some of your sequences have been truncated to be shorter than the
k you're using?

(There's an issue file for this that I can't find at the moment, but no
fix; it's definitely a bug on our part.)

**Comment by bede** [comment]
Ahah! I normally use trimmed/filtered reads but these are virginous, so I bet that's the issue. Will confirm after coffee.

**Comment by bede** [comment]
You were spot on – shortest read was 20bp. Dropped to 19-mers and all works.

Relevant issue:
'Test all scripts/ for appropriate behavior when len(read) < K'
https://github.com/dib-lab/khmer/issues/859 

I've had a look at `utils.py`… It appears a fix might involve passing `min_length` to `check_is_pair()` ?
Sound plausible? I'll struggle more with the test I imagine.

**Comment by bede** [comment]
...I suppose the correct default in paired mode is to ditch both reads if the length of either read is < k

**Comment by mr-c** [comment]
@bede I think keeping data is a better default than silently dropping data though that might be a complicated code path.

**Comment by ctb** [comment]
See fix in #1449, which silently ditches both reads.  While I sympathize with @mr-c's desire to avoid dropping data, we're caught in the following trap:

* require_paired requires that properly paired reads be both consumed and emitted;
* there is no way for many of the scripts to properly reads shorter than K;
* we have no system or provision for recording what happens to odd reads.

The proper thing to do is probably to include this in logging output somewhere, but we also don't have a good system for that :(.
